File: solutions/496.next-greater-element-i/next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int[] res = new int[nums1.length];
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
        Stack<Integer> stack = new Stack<Integer>();
        for(int num : nums2){
            while(!stack.isEmpty() && stack.peek() < num){
                map.put(stack.pop(), num);
            }
            stack.push(num);
        }
        for(int i = 0; i < nums1.length; i++){
            res[i] = map.getOrDefault(nums1[i], -1);
        }
        return res;
    }
}

File: solutions/496.next-greater-element-i/next-greater-element-i.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var nextGreaterElement = function(nums1, nums2) {
    const map = new Map();
    const stack = [];
    const res = [];
    for(let num of nums2){
        while(stack.length > 0 && stack[stack.length - 1] < num){
            map.set(stack.pop(), num);
        }
        stack.push(num);
    }
    for(let i = 0; i < nums1.length; i++){
        if(map.get(nums1[i])){
            res[i] = map.get(nums1[i]);
        }else{
            res[i] = -1;
        }
    }
    return res;
};